Refactor code to implement facade and factory patterns

rainerangelo opened the pull request:
As part of a project for a university course on software architecture, we are creating this pull request which implements a facade and factory pattern.

The facade pattern creates an interface for creating either a geocoding or Foursquare HTTP request.

The factory pattern creates an interface for generating a list of venues using either simple latitude/longitude coordinates or Foursquare API output.

Commit:
Refactor HTTP request code to use facade pattern

- Instead of directly using GeocodingRequest.java or building a
  Foursquare request in RouletteServlet.java, one can simply use a
  RequestMaker object to build either HTTP request

Commit:
Set up factory pattern for venue-creating

Commit:
Add more code for factory pattern

Commit:
Complete factory pattern for FoursquareVenueList

NolanChan1 committed:
Update Location.java

Added setters

NolanChan1 committed:
Added setters

NolanChan1 committed:
added setter

NolanChan1 committed:
Added SimpleVenueList

Commit:
Push to the right directory

Commit:
Delete files from wrong directory

clinejj commented on maproulette/src/main/java/com/pixeltron/mapquest/open/geocoding/RequestMaker.java:
```diff
@@ -0,0 +1,35 @@
+
+    private Request geoReq, fsReq;
+
+    public RequestMaker() {
```
It might be a good option to add a private `RequestParameters` that is set in the constructor, as it would make sense that any request that would be made will require it (and is the case for both the geocoding and foursquare). This would allow for passing in at construction time rather than requiring it to be passed in for each type of request. This would have the `RequestMaker` class be more stateful, rather than simply a proxy object for the underlying request builders.

clinejj commented on maproulette/src/main/java/com/pixeltron/mapquest/open/geocoding/GeocodingRequest.java:
```diff
@@ -5,27 +5,28 @@
 		}
-		if (thumbMaps != null) {
-			urlBuilder.append("&thumbMaps=").append(thumbMaps);
+		if (THUMB_MAPS != null) {
```
This will never be null with the given code changes - I would either update it to read off of parameters (eg `parameters.thumbMaps` or remove the null check. Reading off of `parameters` would be more idiomatic with the rest of the changes.

clinejj commented on maproulette/src/main/java/com/pixeltron/mapquest/open/geocoding/RequestMaker.java:
```diff
@@ -0,0 +1,35 @@
+package com.pixeltron.mapquest.open.geocoding;
+
+public class RequestMaker {
```
The name of this class is potentially a bit misleading - the only interfaces it supports are really just _building_ requests, and not necessarily making them. A more descriptive update would be to make this `RequestBuilder`, or alternately update the `makeRequest` methods to return a `Future<HTTPResponse>` and call the fetch service within this class.

clinejj commented on maproulette/src/main/java/com/pixeltron/mapquest/open/geocoding/RequestParameters.java:
```diff
@@ -0,0 +1,29 @@
+
+import javax.servlet.http.HttpServletRequest;
+
+public class RequestParameters {
```
This is a great change! Makes the parent code much cleaner.

An option could be to make this an interface, and have separate implementations for geocode vs foursquare (similar to the `Request` object). This would make it more consistent from a design pattern (eg I need FoursquareRequestParameters to make a FoursquareRequest) and would allow for having the class only set the necessary parameters.

Additionally, depending on how open or closed you intend this class to be, it might be useful to make these variables `final`. This would prevent issues where certain parameters may get changed unintentionally between uses of the instance.

clinejj reviewed (commented):
Overall these are really great improvements! As mentioned in the comments, I think the main design changes that could be interesting and make this a bit more holistic:

1) Continuing on the design work started by having separate geocode/foursquare requests, also have separate requestparameters that only use the needed fields for that type of request
2) Potentially renaming the `RequestMaker` to `RequestURLBuilder` if keeping the current design the same, or updating it to call the fetch method and actually make the requests.

Commit:
Uncomment SimpleVenueList instantiation

Commit:
Separate geocoding and Foursquare requests

- Created two new classes called GeocodingRequestParameters and
  FoursquareRequestParameters, both of which inherit from the newly
  abstracted RequestParameters
- Changed the name of RequestMaker to RequestBuilder
- Added a private variable for RequestParameters in RequestBuilder,
  which is set in the constructor

rainerangelo commented:
Committed new changes which include improvements based on feedback provided by @clinejj. Main improvement is the separation of geocoding and Foursquare request parameters through the abstraction of the RequestParameters class.

clinejj commented:
New changes look great!
